flink-runtime/src/main/java/org/apache/flink/runtime/leaderelection/ZooKeeperLeaderElectionDriver.java

```diff
@@ -0,0 +1,292 @@
+import org.apache.flink.shaded.zookeeper3.org.apache.zookeeper.data.Stat;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.UUID;
+
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/**
+ * {@link LeaderElectionDriver} implementation for Zookeeper. The leading JobManager is elected using
+ * ZooKeeper. The current leader's address as well as its leader session ID is published via
+ * ZooKeeper.
+ */
+public class ZooKeeperLeaderElectionDriver implements LeaderElectionDriver, LeaderLatchListener, NodeCacheListener, UnhandledErrorListener {
+
+	private static final Logger LOG = LoggerFactory.getLogger(ZooKeeperLeaderElectionDriver.class);
+
+	/** Client to the ZooKeeper quorum. */
+	private final CuratorFramework client;
+
+	/** Curator recipe for leader election. */
+	private final LeaderLatch leaderLatch;
+
+	/** Curator recipe to watch a given ZooKeeper node for changes. */
+	private final NodeCache cache;
+
+	/** ZooKeeper path of the node which stores the current leader information. */
+	private final String leaderPath;
+
+	private final ConnectionStateListener listener = (client, newState) -> handleStateChange(newState);
+
+	private final LeaderElectionEventHandler leaderElectionEventHandler;
+
+	private final String leaderContenderDescription;
+
+	private volatile boolean running;
+
+	/**
+	 * Creates a ZooKeeperLeaderElectionDriver object.
+	 *
+	 * @param client Client which is connected to the ZooKeeper quorum
+	 * @param latchPath ZooKeeper node path for the leader election latch
+	 * @param leaderPath ZooKeeper node path for the node which stores the current leader information
+	 * @param leaderElectionEventHandler event handler for processing leader change events
+	 * @param leaderContenderDescription leader contender description
+	 */
+	public ZooKeeperLeaderElectionDriver(
+			CuratorFramework client,
+			String latchPath,
+			String leaderPath,
+			LeaderElectionEventHandler leaderElectionEventHandler,
+			String leaderContenderDescription) throws Exception {
+		this.client = checkNotNull(client);
+		this.leaderPath = checkNotNull(leaderPath);
+		this.leaderElectionEventHandler = checkNotNull(leaderElectionEventHandler);
+		this.leaderContenderDescription = checkNotNull(leaderContenderDescription);
+
+		leaderLatch = new LeaderLatch(client, checkNotNull(latchPath));
+		cache = new NodeCache(client, leaderPath);
+
+		client.getUnhandledErrorListenable().addListener(this);
+
+		leaderLatch.addListener(this);
+		leaderLatch.start();
+
+		cache.getListenable().addListener(this);
+		cache.start();
+
+		client.getConnectionStateListenable().addListener(listener);
+
+		running = true;
+	}
+
+	@Override
+	public void close() throws Exception{
+		if (!running) {
+			return;
+		}
+		running = false;
```

Comment:
I'd suggest to introduce a `lock` to guard accesses to `running` and all other places where it is accessed. That way we will prevent concurrency when shutting this component down and when the `ZooKeeper` threads trigger an update.